### PR TITLE
Add IAM cred caching for OIDC flow

### PR DESCRIPTION
**Pull request by Manouchehri**
## Title

This should improve the perf, as this allows avoiding making an extra call to STS on every invocation of Bedrock. 

Should be merged after #3688. Doing this as a separate PR in case you don't like my caching method; I don't want this PR to slow down #3688 from being merged.

## Type

🆕 New Feature

## Changes

Just uses DualCache to speed things up!

## [REQUIRED] Testing - Attach a screenshot of any new tests passing locall
If UI changes, send a screenshot/GIF of working UI fixes

I ran the same request multiple times, and could see the cache being used in LiteLLM's logs. (It's kinda difficult to post a screenshot without including a lot of PII.)

image

### CODE REVIEW

File: litellm/llms/bedrock_httpx.py
Comment (ghost) [inline litellm/llms/bedrock_httpx.py:340]: fixing this - https://github.com/BerriAI/litellm/issues/3858, is a precursor to this PR, since this would also introduce sync flows to an async call

**Review by ghost — CHANGES_REQUESTED** [review]
https://github.com/BerriAI/litellm/issues/3858 blocks this pr 

since the refactoring to make the call async will require a refactor here

**Comment by ghost** [comment]
@Manouchehri please can we add some testing for this? couple users use the sts flow and i want to make sure there's no regressions

File: litellm/llms/bedrock_httpx.py
Comment (Manouchehri) [inline litellm/llms/bedrock_httpx.py]: Note for tomorrow David: this need to have `endpoint_url=f"https://sts.{aws_region_name}.amazonaws.com"` as well, STS sorta ignores `region_name` it seems.

**Comment by Manouchehri** [comment]
> @Manouchehri please can we add some testing for this? couple users use the sts flow and i want to make sure there's no regressions

Definitely, that's a good idea, I'll do. We sorta ish already test it, but it's more by luck and not intentional. 

Two questions: 

1. How can I validate that the cache is actually working and not just issuing a new token?
2. How can I allow LiteLLM to only do one STS call at once? There's no point in getting multiple tokens in parallel (it will be the same speed), it's just wasteful on resources and adds noise.

File: litellm/llms/bedrock_httpx.py
Comment (Manouchehri) [reply litellm/llms/bedrock_httpx.py]: Otherwise, boto3 oddly relies on us-east-1. Terrible design decision. https://github.com/boto/boto3/issues/1859#issuecomment-461622548

File: litellm/llms/bedrock_httpx.py
Comment (Manouchehri) [reply litellm/llms/bedrock_httpx.py:340]: Let's tackle this in a different PR, since this is purely focused on the Bedrock IAM part. (While #3858 mentions the host's OIDC token being sync.)

**Comment by Manouchehri** [comment]
This shaves a lot of time off the current setup, between like 0.25 to 1 second per Bedrock call. :) 

The STS fix is also needed.

Ready for merging!

**Comment by Manouchehri** [comment]
Bump on this? It's becoming a bit of a burden to maintain out-of-tree patches for weeks.

**Comment by ghost** [comment]
hey @Manouchehri missed this - sorry for the delay on my end. 

Saw you made a couple changes - could you walk me through it?

**Comment by ghost** [comment]
i'm also happy to hop on a call and help get all the oidc pr's done

https://calendly.com/d/4mp-gd3-k5k/litellm-1-1-onboarding-chat

**Comment by Manouchehri** [comment]
> i'm also happy to hop on a call and help get all the oidc pr's done

Gimme a moment to finish testing this and #3861. =)

**Comment by Manouchehri** [comment]
You free now?

File: litellm/llms/bedrock_httpx.py
Comment (ghost) [inline litellm/llms/bedrock_httpx.py:58]: switch to in memory cache

Comment (ghost) [reply litellm/llms/bedrock_httpx.py:58]: InMemoryCache - https://github.com/BerriAI/litellm/blob/fb96f07ccb28fffbfe7bd9e9723d48c16c4bb581/litellm/caching.py#L59

File: litellm/llms/bedrock_httpx.py
Comment (ghost) [inline litellm/llms/bedrock_httpx.py:370]: can we have ttl be an enum at the top of the file, so it's easier to know

File: litellm/llms/bedrock_httpx.py
Comment (ghost) [inline litellm/llms/bedrock_httpx.py:330]: can you add a comment explaining when you'd enter this flow

**Comment by Manouchehri** [comment]
Self note: I can confirm that this is indeed working, I now only see one `AssumeRoleWithWebIdentity` event before multiple Bedrock calls. =) 

image

Previously (before this PR was merged), I had multiple `AssumeRoleWithWebIdentity` calls (one for each Bedrock invocation, which is waaaay too high of an overhead).

image

**Pull request merged**
